### PR TITLE
Bump mac editor and implement CocoaTextBufferVisibilityTracker

### PR DESCRIPTION
**Pull request by Therzok**
This should help reduce the number classification marks triggering in the editor in VSMac.

Uses newly added API in VS-Platform, 0f93736f636b83a30c3a0b393b89afbe882de5e3.

Fixes https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1689011

### CODE REVIEW

File: src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            var views = _associatedViewService.GetAssociatedTextViews(subjectBuffer).ToImmutableArrayOrEmpty();
+
+            // If we don't have any views at all, then assume the buffer is visible.
+            if (views.Length == 0)
+                return true;
+
+            // if any of the views were *not* wpf text views, assume the buffer is visible.  We don't know how to
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs:75]: "wpf"

Comment (Therzok) [reply src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs:75]: Seems like I missed one :p It's copy pasted from the WPF impl. Want me to do a split with shared code and partial classes? I imagine that makes things simpler.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs:75]: yes.  or just have a generic base type that does all the logic and is parameterized on TTextView, then subclass with the wpf and cocoa impls.

Comment (Therzok) [reply src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs:75]: Yeah, that sounds good. Either works for me.

**Comment by Therzok** [comment]
I tried multiple ways of fixing the duplication, but I end up with an issue where `IsVisibleChanged` is a plain `EventHandler`, and `DependencyPropertyChangedEventArgs` is a struct. Can't wrap the delegates into a compatible one, because there's no conversion that can be done.

Can we take this as-is for now?

File: src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs (hunk 1 above)
Comment (Therzok) [reply src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs:75]:
```suggestion
            // if any of the views were *not* cocoa text views, assume the buffer is visible.  We don't know how to
```

**Comment by CyrusNajmabadi** [comment]
@Therzok Sorry, i should have been more clear on how to do it.  I've made the changed and pushed to your branch.  Let me know how it looks for you.

File: eng/Versions.props [hunk 2]
```diff
@@ -28,7 +28,7 @@
     <VisualStudioEditorPackagesVersion>17.5.42-preview</VisualStudioEditorPackagesVersion>
     <!-- This should generally be set to $(VisualStudioEditorPackagesVersion),
          but sometimes EditorFeatures.Cocoa specifically requires a newer editor build. -->
-    <VisualStudioMacEditorPackagesVersion>17.3.133-preview</VisualStudioMacEditorPackagesVersion>
+    <VisualStudioMacEditorPackagesVersion>17.5.152-preview-g356e9d7141</VisualStudioMacEditorPackagesVersion>
```
Comment (CyrusNajmabadi) [inline eng/Versions.props:31]: @genlu @allisonchou what branch should this go into?  main-vs-deps?

File: src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs [hunk 3]
```diff
@@ -3,223 +3,39 @@
...
-using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Workspaces
 {
     [Export(typeof(ITextBufferVisibilityTracker))]
-    internal sealed class WpfTextBufferVisibilityTracker : ITextBufferVisibilityTracker
+    internal sealed class WpfTextBufferVisibilityTracker
+        : AbstractTextBufferVisibilityTracker<IWpfTextView, DependencyPropertyChangedEventHandler>
     {
-        private readonly ITextBufferAssociatedViewService _associatedViewService;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs:24]: all this moved to abstract base class.

File: src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+namespace Microsoft.CodeAnalysis.Workspaces
+{
+    internal abstract class AbstractTextBufferVisibilityTracker<
+        TTextView,
+        TVisibilityChangedCallback> : ITextBufferVisibilityTracker
+        where TTextView : ITextView
+        where TVisibilityChangedCallback : System.Delegate
+    {
+        private readonly ITextBufferAssociatedViewService _associatedViewService;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs:25]: all this state and logic is the same.  just pulled up into base class.  i have marked where they differ.

File: src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+            _threadingContext = threadingContext;
+
+            associatedViewService.SubjectBuffersConnected += AssociatedViewService_SubjectBuffersConnected;
+            associatedViewService.SubjectBuffersDisconnected += AssociatedViewService_SubjectBuffersDisconnected;
+        }
+
+        protected abstract bool IsVisible(TTextView view);
+        protected abstract TVisibilityChangedCallback GetVisiblityChangeCallback(VisibleTrackerData visibleTrackerData);
+        protected abstract void AddVisibilityChangedCallback(TTextView view, TVisibilityChangedCallback visibilityChangedCallback);
+        protected abstract void RemoveVisibilityChangedCallback(TTextView view, TVisibilityChangedCallback visibilityChangedCallback);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs:44]: abstract functionality for where wpf and cocoa need to specialize.

File: src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+                return true;
+
+            // if any of the views were *not* the right kind of text views, assume the buffer is visible.  We don't know
+            // how to determine the visibility of this buffer.  While unlikely to happen, this is possible with VS's
+            // extensibility model, which allows for a plugin to host an ITextBuffer in their own impl of an ITextView.
+            // For those cases, just assume these buffers are visible.
+            if (views.Any(static v => v is not TTextView))
+                return true;
+
+            return views.OfType<TTextView>().Any(v => IsVisible(v));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs:86]: defer to IsVisible here to do platform specific determination.

File: src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+            /// </summary>
+            public ImmutableHashSet<Action> Callbacks { get; private set; } = ImmutableHashSet<Action>.Empty;
+
+            public VisibleTrackerData(
+                AbstractTextBufferVisibilityTracker<TTextView, TVisibilityChangedCallback> tracker,
+                ITextBuffer subjectBuffer)
+            {
+                _tracker = tracker;
+                _subjectBuffer = subjectBuffer;
+                _visibilityChangedCallback = tracker.GetVisiblityChangeCallback(this);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs:158]: defer to platform specific code to get the callback with which to register to visibility changed event.

File: src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+                {
+                    if (removedView is TTextView genericView)
+                        _tracker.RemoveVisibilityChangedCallback(genericView, _visibilityChangedCallback);
+                }
+
+                // Connect to hearing about visbility changes for any views we are associated with.
+                foreach (var addedView in addedViews)
+                {
+                    if (addedView is TTextView genericView)
+                        _tracker.AddVisibilityChangedCallback(genericView, _visibilityChangedCallback);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs:212]: defer to platform specific code to attach/detach the visibility changed listener.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski ptal.

File: src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs [hunk 9]
```diff
@@ -3,223 +3,37 @@
...
-                UpdateAssociatedViews();
-            }
+        protected override bool IsVisible(ICocoaTextView view)
+            => view.IsVisible;
 
-            public void Dispose()
-            {
-                _tracker._threadingContext.ThrowIfNotOnUIThread();
+        protected override EventHandler GetVisiblityChangeCallback(VisibleTrackerData visibleTrackerData)
+            => (sender, args) => visibleTrackerData.TriggerCallbacks();
```
Comment (Therzok) [inline src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs:31]: Ah, heh, I didn't think of wrapping in the derived impl.

Comment (Therzok) [reply src/EditorFeatures/Core.Cocoa/Workspaces/CocoaTextBufferVisibilityTracker.cs:31]: Thanks for the changes!

**Comment by Therzok** [comment]
> @Therzok Sorry, i should have been more clear on how to do it. I've made the changed and pushed to your branch. Let me know how it looks for you.

No worries! The problem I was having is that I didn't have a plain EventHandler. My WIP work was similar to this, except the GetVisibilityChangeCallback method:
https://gist.github.com/Therzok/29f7a0a73958414008beb909acf121ea

File: eng/Versions.props (hunk 2 above)
Comment (sandyarmstrong) [reply eng/Versions.props:31]: We should be able to bump this to a standard daily build version, yes (won't have a commit hash suffix)?

**Review by sandyarmstrong — CHANGES_REQUESTED** [review]
Looks good, should just probably bump the editor package to a standard build instead of a PR build. 17.5.157-preview is the first that has the necessary changes. 17.5.170-preview is the latest that currently exists.

File: eng/Versions.props (hunk 2 above)
Comment (sandyarmstrong) [reply eng/Versions.props:31]: Yeah, this is a package from a PR build. Let's change this to 17.5.157-preview (or 17.5.170-preview if we want latest).

**Comment by Therzok** [comment]
Yeah, sorry, I was using a PR build to be able to validate the changes first.

File: eng/Versions.props (hunk 2 above)
Comment (sandyarmstrong) [reply eng/Versions.props]: Actually `VisualStudioEditorPackagesVersion` should be bumped and this property's value should be changed to `$(VisualStudioEditorPackagesVersion)`. But that can happen in another PR.

Comment (davidwengier) [reply eng/Versions.props]: Given this (hopefully?) only affects Mac we shouldn't need to worry about vs-deps. If we did bump `VisualStudioEditorPackagesVersion` then we might

**Comment by sandyarmstrong** [comment]
@CyrusNajmabadi @davidwengier can we merge this now?

**Comment by CyrusNajmabadi** [comment]
yup yup!

**Pull request merged**